streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java

```diff
@@ -181,15 +189,16 @@ public void testRolling() {
             ),
             segmentDirs(baseDir)
         );
-
+        // expired record
         assertEquals(
-            new HashSet<>(Collections.singletonList("zero")),
+            new HashSet<>(Collections.emptyList()),
```

Comment:
Could we make the retention period longer to keep the original tests? I'm afraid this change will not catch the expected error when code changes unexpectedly. I'm thinking we can have 2 tests, one is the long retention period, and the other one is current setting.

Reply:
@showuon , can you plz give an example of `I'm afraid this change will not catch the expected error when code changes unexpectedly` ? The reason I ask is, because the underlying fetch logic has changed, it's slightly difficult to find one generic high retention period to make all test cases work as is. I tried by setting a range of values upto 10 * 60k, but still there are test failures.

Also, the reason I changed the test cases is that, the test cases should also reflect the expired records. I agree that ideally we should have been creating 2 separate test cases, but with the changes made, the values that get returned depend upon what has been stored so far which entails test cases change anyways.

Let me know if you agree to this..

Reply:
> I tried by setting a range of values upto 10 * 60k, but still there are test failures.

Could you explain why this will fail tests? I thought it will work as before? Thanks.

Reply:
Sure @showuon .. Let's consider the test case => `shouldOnlyIterateOpenSegments` Now when I run it by setting a very high retention period (like 5 * SEGMENT_INTERVAL ), the test case fails with this assertion => 

```
expected:<KeyValue(60000, two)> but was:<KeyValue(0, one)>
Expected :KeyValue(60000, two)
Actual   :KeyValue(0, one)
```

So, the testcase expects that after the 4th put [here](https://github.com/apache/kafka/pull/11211/files/af361d42295b1f116ecbed6b5721f66b093ba98b#diff-ba861b0f1a8d8f4ae8f2b8464e168c29fc5470cad94ffda1c8fa0b3c7c15b72bR135), it should expire the older segment and return only 2/3 from the fetch. I looked at the values further and retention is 300k and segmentInterval is 150k for all stores. Now, the 4th put does lead to the creation of a new segment (as it's 180k/150k) but the older segment doesn't still get expired as the observedStreamTime (180k) < retentionPeriod(300k) which means it returns the first put as well. 

I think this is how the store iterators work currently and the changes in this PR are not impacting this behaviour. With my change, for TimeOrderedWindowStores, this comment should explain why I needed to change the test case:

```
// In case of RocksDBTimeOrderedWindowStoreWithIndex, there's an extra get call to check if a key exists in base store or not
                // now, that record is expired so as it's ts is 60,000 which is < (observedTime - retention + 1) == 60,001
                // and hence that record is also not returned.
```

I think to have the test cases work as is with a high retention, I will need to change/tweak the segment_interval value to have the retentionPeriod kicking in. Currently, the effective segmentInterval set at a store level is defined by =>

`Math.max(retentionPeriod/2, segmentInterval)` and I think that's the reason in the test case initialization, retention is set to 2 * SEGMENT_INTERVAL which neatly fits into the test case. That's because at observedSteamTime = 180k, retention is 120k segment 3 gets rolled out and all segments prior to TS 60k (180k - 120k) gets dropped which passes the test case. 

Other test cases would also follow suit and of-course the expiry of records from this PR would also kick in.

Reply:
@showuon , did you get a chance to review this comment? Thanks!

Reply:
@vamossagar12 , thanks for the explanation! Make sense to me.